### PR TITLE
cmd/swarm: add swarm down

### LINKED ISSUE #439: Add a `swarm down` command that is dual to `swarm up`
I suggest we add `swarm down` to the CLI interface.

It can be just a simple wrapper for `curl`.

### PR DESCRIPTION
**Pull request by acud**
fixes #439

### CODE REVIEW

File: cmd/swarm/download.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		}
+
+		log.Debug(fmt.Sprintf("swarm download: downloading file/path from a manifest. hash: %s, path:%s", uri.Addr, uri.Path))
+
+		file, err := client.Download(uri.Addr, uri.Path)
+		if err != nil {
+			utils.Fatalf("could not download %s from given address: %s. error: %v", uri.Path, uri.Addr, err)
+		}
+
+		re := regexp.MustCompile("[^/]+$") //everything after last slash
```
Comment (zelig) [inline cmd/swarm/download.go]: what is this logic after this line i dont get it.
If download is not handled by the client call, it should be fixed there no?

File: cmd/swarm/download.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/swarm/api"
+	swarm "github.com/ethereum/go-ethereum/swarm/api/client"
+	"gopkg.in/urfave/cli.v1"
+)
+
+func download(ctx *cli.Context) {
+	log.Debug("swarm down")
```
Comment (lmars) [inline cmd/swarm/download.go]: This isn't very informative.

File: cmd/swarm/download.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		bzzapi      = strings.TrimRight(ctx.GlobalString(SwarmApiFlag.Name), "/")
+		isRecursive = ctx.GlobalBool(SwarmRecursiveUploadFlag.Name)
+		client      = swarm.NewClient(bzzapi)
+	)
+
+	dir := ""
+
+	if len(args) == 1 {
+		// no destination arg - assume current terminal working dir
+		workingDir, err := filepath.Abs("./")
```
Comment (lmars) [inline cmd/swarm/download.go]: Just initialise `dir := "."` above, there is no need to get an absolute path.

File: cmd/swarm/download.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	if err != nil {
+		utils.Fatalf("could not stat path")
+	}
+
+	switch mode := fi.Mode(); {
+	case mode.IsRegular():
+		utils.Fatalf("destination path is not a directory!")
+	}
+
+	uri, err := api.Parse(args[0])
```
Comment (lmars) [inline cmd/swarm/download.go]: Check this error.

File: cmd/swarm/download.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+func download(ctx *cli.Context) {
+	log.Debug("swarm down")
+	args := ctx.Args()
+	if len(args) < 1 {
+		utils.Fatalf("Usage: swarm down <bzz locator> [<destination path>]")
+	}
+
+	var (
+		bzzapi      = strings.TrimRight(ctx.GlobalString(SwarmApiFlag.Name), "/")
+		isRecursive = ctx.GlobalBool(SwarmRecursiveUploadFlag.Name)
```
Comment (lmars) [inline cmd/swarm/download.go]: This should not be a global flag.

File: cmd/swarm/main.go [hunk 6]
```diff
@@ -240,6 +240,18 @@ Lists files and directories contained in a manifest.
 Prints the swarm hash of file or directory.
 `,
 		},
+		{
+			Action: download,
+			Name:   "down",
+			Flags:  []cli.Flag{SwarmRecursiveUploadFlag},
+
+			Usage:     "downloads a swarm manifest or a file inside a manifest",
+			ArgsUsage: " <bzzhash> [<download path>]",
```
Comment (lmars) [inline cmd/swarm/main.go]: The arguments should be `<uri> [<dir>]` with an explanation of each.

File: cmd/swarm/upload_test.go [hunk 7]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
 		if string(reply) != data {
 			t.Fatalf("expected HTTP body %q, got %q", data, reply)
 		}
+		log.Info("verifying uploaded file using `swarm down`")
+		//try to get the content with `swarm down`
+		tmpDownload, err := ioutil.TempDir("", "swarm-test")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.Remove(tmpDownload)
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: This should call `os.RemoveAll`

File: cmd/swarm/upload_test.go [hunk 8]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+
+		bzzLocator := "bzz:/" + hash
+		flagss := []string{}
+		flagss = []string{
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: This should just be `flags := []string{ ... }`

File: cmd/swarm/upload_test.go [hunk 9]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+			bzzLocator,
+			tmpDownload,
+		}
+		if toEncrypt {
+			flagss = []string{
+				"--verbosity", "5",
+				"--bzzapi", cluster.Nodes[0].URL,
+				"down",
+				bzzLocator,
+				tmpDownload}
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: This list of flags is the same, should it include `--encrypt`?

File: cmd/swarm/upload_test.go [hunk 10]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+		}
+
+		switch mode := fi.Mode(); {
+		case mode.IsRegular():
+			if file, err := swarm.Open(path.Join(tmpDownload, hash)); err != nil {
+				t.Fatalf("encountered an error opening the file returned from the CLI: %v", err)
+			} else {
+
+				ff := make([]byte, len(data))
+				io.ReadFull(file, ff)
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: Use [`ioutil.ReadFile`](https://golang.org/pkg/io/ioutil/#ReadFile)

File: cmd/swarm/upload_test.go [hunk 11]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+				buf := bytes.NewBufferString(data)
+
+				if !bytes.Equal(ff, buf.Bytes()) {
+					t.Fatalf("retrieved data and posted data not equal!")
+				}
+			}
+		default:
+			t.Fatalf("this shouldnt happen")
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: This error should be more informative, for example:

```go
t.Fatalf("expected to download regular file, got %s", fi.Mode())
```

File: cmd/swarm/upload_test.go [hunk 12]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+	httpClient := http.Client{
+		Timeout: timeout,
 	}
 
-	// get an non-existent hash from each node
+	// try to squeeze a timeout by getting an non-existent hash from each node
 	for _, node := range cluster.Nodes {
-		res, err := http.Get(node.URL + "/bzz:/1023e8bae0f70be7d7b5f74343088ba408a218254391490c85ae16278e230340")
+		_, err := httpClient.Get(node.URL + "/bzz:/1023e8bae0f70be7d7b5f74343088ba408a218254391490c85ae16278e230340")
+		// we're speeding up the timeout here since netstore has a 60 seconds timeout on a request
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: There should be a request header / query parameter (e.g. `GET /bzz:/<hash>/?timeout=1s`), although this seems unrelated to the addition of `swarm down`?

File: cmd/swarm/upload_test.go [hunk 13]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tmp2.Close()
+	defer os.Remove(tmp1.Name())
+	defer os.Remove(tmp1.Name())
+
+	// write data to file
+	data := "notsorandomdata"
+	_, err = io.WriteString(tmp1, data)
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: Rather than using `ioutil.TempFile` and `io.WriteString`, the temp file creation can just be:

```go
for _, path := range []string{"tmp1", "tmp2"} {
	if err := ioutil.WriteFile(filepath.Join(tmpDir, path), data, 0644); err != nil {
		t.Fatal(err)
	}
}
```

File: cmd/swarm/upload_test.go [hunk 14]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if res.StatusCode != 200 {
+			t.Fatalf("expected HTTP status 200, got %s", res.Status)
+		}
+
+		_, err = ioutil.ReadAll(res.Body)
+		defer res.Body.Close()
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: This should be deferred straight after the err check from `http.Get`

File: cmd/swarm/upload_test.go [hunk 15]
```diff
@@ -101,16 +121,207 @@ func testCLISwarmUp(toEncrypt bool, t *testing.T) {
...
+		res, err := http.Get(node.URL + "/bzz-list:/" + hash)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if res.StatusCode != 200 {
+			t.Fatalf("expected HTTP status 200, got %s", res.Status)
+		}
+
+		_, err = ioutil.ReadAll(res.Body)
```
Comment (lmars) [inline cmd/swarm/upload_test.go]: Why does the test call `bzz-list://<hash>` but then throw away the result?

File: swarm/api/client/client.go [hunk 16]
```diff
@@ -235,6 +238,96 @@ func (c *Client) DownloadDirectory(hash, path, destDir string) error {
...
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected HTTP status: %s", res.Status)
+	}
+	tr := tar.NewReader(res.Body)
```
Comment (lmars) [inline swarm/api/client/client.go]: Given this method is called `DownloadFile`, why is it downloading a tar archive (which is what `DownloadDirectory` is for)?

**Comment by lmars** [comment]
@justelad 

I'm not sure only allowing a directory as an argument is useful, given the most common case will probably be downloading a single file.

I think the usage should be something like this:

```
# download a file to file.txt
$ swarm down bzz://mysite.eth/file.txt file.txt

# download a file into dir/
$ swarm down bzz://mysite.eth/file.txt dir/

# download a directory into dir/
$ swarm down --recursive bzz://mysite.eth/files/ dir/
```

File: cmd/swarm/upload_test.go (hunk 12 above)
Comment (acud) [reply cmd/swarm/upload_test.go]: unrelated. this is due to the fact that the netstore timeout is set to 60 seconds. the solution in this case would be to pull the timeout in the netstore to a config flag, then we could instantiate the tests with lower timeouts (and then i'd get rid of this timeout). not sure if timeout should be a querystring param in this case. it's non-conventional. i'd rather let the transport layer to deal with timeouts (or specific built in timeouts)

**Comment by acud** [comment]
I used the tar stream option due to the fact it supports for a few cases in one go - since the content can either be a direct content hash (that does not has a notion of filename), and if the file sits in a manifest, it gets returned with that information in 1 request instead of 2 different ones, traversing the manifest (or detecting its a single file), and asserting different filename possibilities.
ideally we should return a `Content-Disposition` header from the HTTP server and have browsers/clients assert the filename through that.

**Comment by lmars** [comment]
@justelad 

> ideally we should return a Content-Disposition header from the HTTP server and have browsers/clients assert the filename through that.

I'm not sure I understand why that is necessary.

If I run `swarm down bzz://mysite.eth/file.txt` then the filename can be determined from the URL (i.e. it's `file.txt`).

This is actually the only way it will work, because the swarm server unfortunately has no concept of a "file".

For example, given two URLs:

* `bzz://mysite.eth/file1.txt`
* `bzz://mysite.eth/file2.txt`

This will be represented by a manifest at `file` with two entries `1.txt` and `2.txt`, so if I run `swarm down bzz://mysite.eth/file1.txt` it's going to return the path as `1.txt`, not `file1.txt`.

I have lobbied many times to adopt the Unix filesystem model to make this more intuitive (initial proposal [here](https://github.com/ethereum/go-ethereum/issues/14349)), but there is little uptake from the team 😞

**Comment by acud** [comment]
@lmars what about a case where a user:
`swarm down bzz:/<hash>` where `hash` points to a manifest with a single entry and no `path` value? it will resolve the content directly without any filename indication?

File: cmd/swarm/download.go (hunk 3 above)
Comment (acud) [reply cmd/swarm/download.go]: what's the argument against absolute paths?

File: cmd/swarm/upload_test.go (hunk 9 above)
Comment (acud) [reply cmd/swarm/upload_test.go]: no, why should it? encrypt is transparent using double-length references

**Comment by lmars** [comment]
> what about a case where a user:
> `swarm down bzz:/<hash>` where hash points to a manifest with a single entry and no path value? it will resolve the content directly without any filename indication?

I would want to run:

```
$ swarm down bzz://<hash> file.txt
```

which would download into `file.txt`.

If I do:

```
$ swarm down bzz://<hash> dir
```

Then it should fail immediately because the destination is a directory and it can't determine the file path from the URL, it would need to be:

```
$ swarm down bzz://<hash> dir/file.txt
```

or if we are trying to download recursively:

```
$ swarm down --recursive bzz://<hash> dir
```

File: cmd/swarm/download.go (hunk 3 above)
Comment (lmars) [reply cmd/swarm/download.go]: Because it is unnecessary, the OS knows that `"."` refers to the current directory, you don't need to explicitly turn it into an absolute path.

File: cmd/swarm/upload_test.go (hunk 9 above)
Comment (lmars) [reply cmd/swarm/upload_test.go]: Ok, so if it doesn't need an extra flag, then why have this `if toEncrypt` that just sets the exact same flags?

Comment (acud) [reply cmd/swarm/upload_test.go]: correct. pushed an update

**Comment by cobordism** [comment]
What happens when I do `swarm down hash-of-manifest /path/to/dir/` and this manifest contains only a default entry?

and what does `swarm down --recursive bzz://<hash> dir` do with default entries in manifests?

-just curious.

**Comment by lmars** [comment]
@homotopycolimit 

> What happens when I do `swarm down hash-of-manifest /path/to/dir/` and this manifest contains only a default entry?

I can't speak for the current code, but I think it should be an error because you are trying to download a file into a directory without giving an indication of what the filename should be. The command should either be:

* `swarm down bzz://<hash> /path/to/dir/file.txt`
* `swarm down --recursive bzz://<hash> /path/to/dir/`

Now, the second command should actually fail too (or not actually download anything) because you can't download the empty path file to `/path/to/dir/` because it is a directory.

> and what does `swarm down --recursive bzz://<hash> dir` do with default entries in manifests?

It will have to fail or just not download paths which end in `/` (e.g. if `dir/` is a manifest with an empty path, that can't map to a file on a Unix filesystem).

File: swarm/api/client/client.go [hunk 17]
```diff
@@ -235,6 +236,92 @@ func (c *Client) DownloadDirectory(hash, path, destDir string) error {
...
+	piper, pipew := io.Pipe()
+	go func() {
+		defer pipew.Close()
+		io.Copy(pipew, res.Body)
+	}()
+
+	io.Copy(dst, piper)
+	piper.Close()
+
+	dst.Close()
```
[the remark this answers is not in the document]
Comment (justelad) [reply swarm/api/client/client.go]: why not use the streams given in the http response and write it directly to the stream of the output file? classic use case of piping streams, which will end on EOF automatically.

[the remark this answers is not in the document]
Comment (lmars) [reply swarm/api/client/client.go]: Oh sorry, I missed that this was writing `res.Body`. There is no need for `io.Pipe` here, it should just be:

```go
dst, err := os.Create(filename)
if err != nil {
	return err
}
defer dst.Close()
_, err = io.Copy(dst, res.Body)
return err
```

You'll notice I've used `os.Create` which fails if the destination already exists. We shouldn't overwrite existing files unless the user explicitly requests it (either by showing a prompt or having a `--force` flag).

**Review by zelig — APPROVED** [review]

**Comment by lmars** [comment]
@justelad downloading into a directory doesn't seem to work:


* create a manifest containing `file.txt`

```
$ curl -F 'file.txt=data;type=text/plain' http://localhost:8500/bzz:/
15bf8ac4b135fc9010d3db40883cd66433849713f5124e56dc821584f4e1125c

$ swarm ls 15bf8ac4b135fc9010d3db40883cd66433849713f5124e56dc821584f4e1125c
HASH                                                              CONTENT TYPE  PATH
9638c5d3288affe342f8edcd72afe13956a7b758dd86a35e5f8307eeade4240b  text/plain    file.txt
```

* create a directory `dir/`

```
$ mkdir dir
$ tree .
.
└── dir

1 directory, 0 files
```

* download into `dir/`

```
$ swarm down bzz://15bf8ac4b135fc9010d3db40883cd66433849713f5124e56dc821584f4e1125c/file.txt dir/

$ tree .
.
├── dir
└── file.txt

1 directory, 1 file
```

As you can see, the file downloaded into the current directory, not `dir/`.

**Comment by acud** [comment]
@lmars should be fixed now

File: swarm/api/client/client.go [hunk 18]
```diff
@@ -235,6 +236,91 @@ func (c *Client) DownloadDirectory(hash, path, destDir string) error {
...
+			}
+		}
+		filename = filepath.Join(dest, filename)
+	}
+	filePath, err := filepath.Abs(filename)
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
```
Comment (janos) [inline swarm/api/client/client.go]: MkdirAll perm argument is before umask, so it is safe to be 0777, allowing users to ser more permissive permissions with umask. In case 0755, users will not be able to set group or other write permissions using umask.

File: cmd/swarm/download.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	uri, err := api.Parse(args[0])
+	if err != nil {
+		utils.Fatalf("could not parse uri argument: %v", err)
+	}
+
+	// assume behaviour according to --recursive switch
+	if isRecursive {
+		if err := client.DownloadDirectory(uri.Addr, uri.Path, dest); err != nil {
```
Comment (janos) [inline cmd/swarm/download.go:73]: If the directory provided as the command argument does not exist, this function returns an error. Do you think that we could create that directory in this case? User probably expect that.

This command:
```
swarm down --recursive bzz://fc652159b778d36168c8f2cb1e450cb59d2a2742737a2f716f01f2246b7c0145 test
```

fails if `test` does not exist, and users must execute `mkdir test` and then the `swarm down` again. Creating the directory may be helpful.

Comment (acud) [reply cmd/swarm/download.go:73]: i rather not support this scenario. i think that its normal behaviour the actually target directories don't get created (from my experience, anyway), while single target files do

Comment (janos) [reply cmd/swarm/download.go:73]: OK, that is fine, too.

**Review by janos — APPROVED** [review]
